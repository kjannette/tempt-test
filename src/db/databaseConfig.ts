export const databaseConfig = {
    connectionString: process.env.DATABASE_CONNECTION_STRING!
};