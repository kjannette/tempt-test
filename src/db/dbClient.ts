import { drizzle } from 'drizzle-orm/node-postgres';
import { Pool } from 'pg';
import { quotes } from './schema/quotes';
import { underwritingDetails } from './schema/underwritingDetails';
import { databaseConfig } from './databaseConfig';

const pool = new Pool(databaseConfig);

export const schema = {
    quotes,
    underwritingDetails
};

export const dbClient = drizzle(pool, {
    schema,
});
