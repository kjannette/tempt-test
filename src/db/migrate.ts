import 'dotenv/config';
import path from 'path';

import { drizzle } from 'drizzle-orm/node-postgres';
import { migrate } from 'drizzle-orm/node-postgres/migrator';
import { Pool } from 'pg';

import { databaseConfig } from './databaseConfig';
import { schema } from './dbClient';

export const migrateDb = async () => {
    const pool = new Pool({
        ...databaseConfig,
        max: 1, // Drizzle recommend setting max to 1 for migrations see https://orm.drizzle.team/docs/get-started-postgresql#node-postgres
    });

    const db = drizzle(pool, { schema });
    await migrate(db, {
        migrationsFolder: path.resolve(__dirname, './migrations'),
        migrationsTable: 'drizzle_migrations',
    });

    await pool.end();
};
