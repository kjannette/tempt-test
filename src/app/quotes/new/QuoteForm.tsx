"use client";


import { createQuote } from "./actions";
import { Path, useForm } from "react-hook-form";
import { useTransition } from "react";

type QuoteFormProps = {};

export type QuoteFormValues = {
    firstName: string;
    lastName: string;
    email: string;
    numberCars: number;
    addressLine1: string;
    addressLine2?: string;
    city: string;
    state: string;
    zip: string;
};

export type NewQuoteFormState = {
    message: string;
    errors: Partial<Record<Path<QuoteFormValues>, string>>;
};

export const QuoteForm = ({}: QuoteFormProps) => {
    const [isPending, startTransition] = useTransition();
  const { register, handleSubmit, formState: { errors } } = useForm<QuoteFormValues>({
    reValidateMode: 'onChange'
  });

  const onSubmit = handleSubmit(data => {
    startTransition(() => {
      createQuote(data as unknown as QuoteFormValues);
    });
  });

  return (
    <form onSubmit={onSubmit}>
    <div className="form-field">
        <label htmlFor="firstName">First Name</label>
        <input
        {...register("firstName", { required: "This field is required" })}
        id="firstName"
        />
        {errors.firstName && <span>{errors.firstName.message}</span>}
    </div>

    <div className="form-field">
        <label htmlFor="lastName">Last Name</label>
        <input
        {...register("lastName", { required: "This field is required" })}
        id="lastName"
        />
        {errors.lastName && <span>{errors.lastName.message}</span>}
    </div>

    <div className="form-field">
        <label htmlFor="email">Email</label>
        <input
        type="email"
        {...register("email", { required: "This field is required" })}
        id="email"
        />
        {errors.email && <span>{errors.email.message}</span>}
    </div>

    <div className="form-field">
        <label htmlFor="numberCars">Number of Cars</label>
        <input
        type="number"
        {...register("numberCars", { required: "This field is required" })}
        id="numberCars"
        />
        {errors.numberCars && <span>{errors.numberCars.message}</span>}
    </div>

    <div className="form-field">
        <label htmlFor="addressLine1">Address Line 1</label>
        <input
        {...register("addressLine1", {
            required: "This field is required",
        })}
        id="addressLine1"
        />
        {errors.addressLine1 && <span>{errors.addressLine1.message}</span>}
    </div>

    <div className="form-field">
        <label htmlFor="addressLine2">Address Line 2</label>
        <input {...register("addressLine2")} id="addressLine2" />
    </div>

    <div className="form-field">
        <label htmlFor="city">City</label>
        <input
        {...register("city", { required: "This field is required" })}
        id="city"
        />
        {errors.city && <span>{errors.city.message}</span>}
    </div>

    <div className="form-field">
        <label htmlFor="state">State</label>
        <input
        {...register("state", { required: "This field is required" })}
        id="state"
        />
        {errors.state && <span>{errors.state.message}</span>}
    </div>

    <div className="form-field">
        <label htmlFor="zip">ZIP Code</label>
        <input
        {...register("zip", { required: "This field is required" })}
        id="zip"
        />
        {errors.zip && <span>{errors.zip.message}</span>}
    </div>
    
    <button type="submit" disabled={isPending}>Submit</button>
    </form>
  );
};
