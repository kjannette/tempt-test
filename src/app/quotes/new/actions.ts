'use server';

import { UnderwritingDetails } from '@/db/schema/underwritingDetails';
import {dbClient, schema} from '../../../db/dbClient';
import { QuoteFormValues } from "./QuoteForm";
import { redirect } from 'next/navigation';
import logger from '@/logger';
import { Quote } from '@/db/schema/quotes';

const calculatePremium = (underwritingDetails: UnderwritingDetails) => +underwritingDetails.numberCars * 50;
const calculateLimit = (underwritingDetails: UnderwritingDetails) => Math.min(+underwritingDetails.numberCars * 10000, 1e6);
const calculateDeductible = () => 5000;

export async function createQuote(formData: QuoteFormValues) { 
    let quote: Quote;

    try {
        const [underwritingDetails] = await dbClient.insert(schema.underwritingDetails).values({
            firstName: formData.firstName,
            lastName: formData.lastName,
            addressLine1: formData.addressLine1,
            addressLine2: formData.addressLine2,
            city: formData.city,
            state: formData.state,
            zip: formData.zip,
            email: formData.email,
            numberCars: formData.numberCars
        }).returning();
    
        [quote] = await dbClient.insert(schema.quotes).values({
            underwritingDetailsId: underwritingDetails.id,
            premium: calculatePremium(underwritingDetails),
            limit: calculateLimit(underwritingDetails),
            deductible: calculateDeductible()
        }).returning();
    } catch (error) {
        logger.error({ event: 'QUOTE:CREATED:ERROR', message: (error as Error).message });
        throw error;
    }

    logger.info({ event: 'QUOTE:CREATED', quoteId: quote.id });

    return redirect(`/quotes/${quote.id}/payment`);
  }
  