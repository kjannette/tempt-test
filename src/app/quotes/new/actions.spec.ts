import { dbClient, schema } from '@/db/dbClient';
import { QuoteFormValues } from './QuoteForm';
import {createQuote} from './actions';

describe('Server action: createQuote', () => {
    it('should persist underwriting details', async () => {
        const formData: QuoteFormValues = {
            firstName: 'John',
            lastName: 'Doe',
            email: 'john.doe@example.com',
            numberCars: 2,
            addressLine1: '123 Main St',
            addressLine2: '',
            city: 'Springfield',
            state: 'IL',
            zip: '62701',
        };

        try {
            await createQuote(formData);
        } catch (err) {
            // Catch the redirect
        }

        const underwritingDetails = await dbClient.select().from(schema.underwritingDetails);
        expect(underwritingDetails).toMatchSnapshot([{
            id: expect.any(String),
            createdAt: expect.any(Date),
            updatedAt: expect.any(Date),
        }]);
    });

    it('should persist a quote', async () => {
        const formData: QuoteFormValues = {
            firstName: 'John',
            lastName: 'Doe',
            email: 'john.doe@example.com',
            numberCars: 2,
            addressLine1: '123 Main St',
            addressLine2: '',
            city: 'Springfield',
            state: 'IL',
            zip: '62701',
        };

        try {
            await createQuote(formData);
        } catch (err) {
            // Catch the redirect
        }

        const quote = await dbClient.select().from(schema.quotes);
        expect(quote).toMatchSnapshot([{
            id: expect.any(String),
            underwritingDetailsId: expect.any(String),
            createdAt: expect.any(Date),
            updatedAt: expect.any(Date),
        }]);
    });

    it('should redirect to the payment page', async () => {
        const formData: QuoteFormValues = {
            firstName: 'John',
            lastName: 'Doe',
            email: 'john.doe@example.com',
            numberCars: 2,
            addressLine1: '123 Main St',
            addressLine2: '',
            city: 'Springfield',
            state: 'IL',
            zip: '62701',
        };

        try {
            await createQuote(formData);
        } catch (err) {
            return;
        }

        expect(1).toBe(2);
    })
});
