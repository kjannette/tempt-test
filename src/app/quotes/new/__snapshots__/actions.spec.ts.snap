// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Server action: createQuote > should persist a quote 1`] = `
[
  {
    "createdAt": Any<Date>,
    "deductible": 5000,
    "id": Any<String>,
    "limit": 20000,
    "premium": 100,
    "underwritingDetailsId": Any<String>,
    "updatedAt": Any<Date>,
  },
]
`;

exports[`Server action: createQuote > should persist underwriting details 1`] = `
[
  {
    "addressLine1": "123 Main St",
    "addressLine2": "",
    "city": "Springfield",
    "createdAt": Any<Date>,
    "email": "john.doe@example.com",
    "firstName": "John",
    "id": Any<String>,
    "lastName": "Doe",
    "numberCars": 2,
    "state": "IL",
    "updatedAt": Any<Date>,
    "zip": "62701",
  },
]
`;
